### ISSUE TITLE
Missing dependencies

### ISSUE THREAD

**Issue by RodricBr** [Kalmat/DSTIKE-SPACEHUHN-Deauther-Weather#1]
While building the project I encountered a lot of missing dependencies that were not included into the directory.
Found some others error like this one: `Compilation error: 'myClock' was not declared in this scope; did you mean 'clock'?`

In general, I could not build into my hardware due to errors.
I'm not a C++ programmer so this is super confusing to me. Some help would be appreciated!

**Comment by Kalmat** [comment]
Hi! Thank you for your interest in this project!

Unfortunately, I am not a C++ programmer either... I had to tweak my installation a lot until I eventually made it work, but not sure exactly how.

The dependencies are installed within Arduino IDE, not sure if they can be installed from outside (I think it is not possible, but I don't really know). 

e.g. 'myClock' variable is declared using one of these dependencies, and will fail if not using the exact proper one.

Let me check which modules I have installed in my Arduino IDE and come back to you with an extensive list.

**Comment by Kalmat** [comment]
Ok. Let's try to put everything in place (this will also help me to improve the installation guide for other users).

1. I am using Arduino IDE 2.2.1
2. Find attached some screenshots showing the modules I have installed (check/manage them in Arduino IDE -> Tools -> Manage Librarties... option). Some libraries will already be installed out-of-the-box, so install only the missing ones. Be sure to check and match the versions too.
3. I am also attaching my last code (it should be the same you can find in the github .zip, but just in case)

Libraries:
![Arduino_Modules_1](https://github.com/user-attachments/assets/5c022c13-71de-4a62-b88c-64135d941e4b)
![Arduino_Modules_2](https://github.com/user-attachments/assets/34bdfbf4-b7a2-472f-ba27-ee32666580ad)
![Arduino_Modules_3](https://github.com/user-attachments/assets/c47c55f9-deae-4c74-90b9-f760b61b40cc)
![Arduino_Modules_4](https://github.com/user-attachments/assets/35c3741c-aea0-453a-84d1-f255dc242551)

Code:
[esp8266_deauther.zip](https://github.com/user-attachments/files/17015109/esp8266_deauther.zip)

You will have to build your own WifiKeys.h file, including your own WiFi keys and your OpenWeatherMap key. It should look like this, but with proper content (there are two wifi SSID/keys sets, one for your home wifi, and the other one for mobility wifi... e.g. your phone WiFi AP):
![WifiKeys](https://github.com/user-attachments/assets/1ddd686b-fa83-4cdf-be15-808cc04ad681)

So, to complete the installation you should:
1. Install all modules within Arduino IDE
2. Unzip code in your folder of choice
3. Create, fill and copy WifiKeys.h within the folder you unzipped the code
4. Double-click esp8266_deauther.ino file. The project should automatically open in Arduino IDE
5. Try to compile

**Comment by RodricBr** [comment]
Ooh thanks! It compiled! Just had to steak a few things and it compiled... but my DSTIKE screen is just blank.
I turned it on and the LED flashed quickly but now there's only a blue and red light shinnying with an empty display.
I pressed the reset button once to see if anything happens but nothing different other than the LED flashing again.

Additional information: I'm on Windows 11 version 23H2. My Arduino IDE is on version 2.3.3 and my DSTIKE is the same shown in the repository, I believe its full name is "DSTIKE Deauther Watch v3".

**Comment by Kalmat** [comment]
Don't worry. I think that the original software is still available at SpaceHuhn github repo if everything goes wrong.

I think to remember you had to install additional software to have it working with the DEAUTHER hardware (sorry, it's a long time I finished this project). Follow the instructions [here](https://deauther.com/docs/diy/installation-arduino) to install the required drivers and other stuff.

In addition to that, you have to select proper values in "Tools" menu option before uploading the code to your watch (it is actually the V3 version, right?).

The values in the README of this repo are taken from the old Arduino IDE version I initially used in my project:
![Configuration](https://github.com/user-attachments/assets/2ad5ab13-0d21-477c-8824-038f0c251e2e)

These are the updated values for Arduino IDE 2.2.1:
![New_Configuration](https://github.com/user-attachments/assets/acde1c96-829d-411b-a44e-7e57695d454b)

If it doesn't work at first try, perhaps you have to "play" with different values which may work in your especific case. For instance, check the differences between "Erase Flash" values. In your case, I would first try using "Scketch only" (which is less aggressive since it only uploads the new program), and then "All Flash Contents" (which may reinstall proper hardware-related software if previous upload you did somehow broke it).

Please, let me know if this works or you find any other issue.

**Comment by RodricBr** [comment]
It worked! Tweaked a few more things in the code again and I had been selecting the wrong board, which was `DSTIKE Deauther` instead of `Generic ESP8266 Module`.
Since I was really invested in testing your adaptation of the Weather Station I can finally say that it looks and works really good, you did a great job!

Here's a picture of it:
<img src="https://github.com/user-attachments/assets/3948057f-fbae-4a49-9edf-be416a41d1e9" width="30%">


Never worked with arduino before so it was painfully new to me. Living and learning.
Thank you so much for your help!

Only one last thing that I noticed and been trying to fix. At **WeatherStation.cpp** line **277** is the place on which the WD (working day) is outputted in the center of the screen. And for me, it is being printed as `?` and the date  as `09/03/200`. The final output would be something like: `?, 09/03/200`
Any idea on why would this be happening?

**Comment by Kalmat** [comment]
Hey! So happy to hear that!!!!

This was also my first (and last, by the moment) project in C++, as well as in Arduinto. It was hard, in deed, but the satisfaction when you make it is enormous! Fully agree: Living is learning.

Regarding the date format, I remember it was ridiculously difficult... This issue also happened to me, but very likely I forgot to upload this last version fixing it. Please replace all these methods within `WeatherStation.cpp` (around line 224, I think):

```
void getTimeRTC() {
  clockHour = myClock.getHour(h12, PM_time);
  clockMinute = myClock.getMinute();
  clockSecond = myClock.getSecond();
  clockDoW = myClock.getDoW();
  clockDate = myClock.getDate();
  clockMonth = myClock.getMonth(century);
  clockYear = myClock.getYear() + 2000;
}

void getTimeNTP() {
  now = time(nullptr);
  timeInfo = localtime(&now);

  clockHour = timeInfo->tm_hour;
  clockMinute = timeInfo->tm_min;
  clockSecond = timeInfo->tm_sec;
  clockDoW = timeInfo->tm_wday;
  clockDate = timeInfo->tm_mday;
  clockMonth = timeInfo->tm_mon + 1;
  clockYear = timeInfo->tm_year + 1900;
}

void setTimeRTCfromNTP() {
  getTimeNTP();
  if (clockYear > 1970) {
    myClock.setHour(clockHour);
    myClock.setMinute(clockMinute);
    myClock.setSecond(clockSecond);
    myClock.setDate(clockDate);
    myClock.setMonth(clockMonth);
    myClock.setYear(clockYear - 2000);
    myClock.setDoW(clockDoW);
  }
}

void getTime() {
  #ifdef RTC_DS3231
    getTimeRTC();
  #else // #ifdef RTC_DS3231
    getTimeNTP();
  #endif // #ifdef RTC_DS3231
}

void drawDateTime(OLEDDisplay *display, OLEDDisplayUiState* state, int16_t x, int16_t y) {

  char buff[16];

  getTime();
 
  display->setTextAlignment(TEXT_ALIGN_CENTER);
  display->setFont(ArialMT_Plain_10);

  sprintf_P(buff, PSTR("%s, %02d/%02d/%04d"), WDAY_NAMES[clockDoW].c_str(), clockDate, clockMonth, clockYear);
  display->drawString(64 + x, 5 + y, String(buff));
  display->setFont(ArialMT_Plain_24);

  sprintf_P(buff, PSTR("%02d:%02d:%02d"), clockHour, clockMinute, clockSecond);
  display->drawString(64 + x, 15 + y, String(buff));
  display->setTextAlignment(TEXT_ALIGN_LEFT);
}
```

If you find issues compiling or the wrong date format remains, try to upload the .zip I attached in my previous reply (it should work OK, just adding WifiKeys.h file).

If it works one way or the other, please let me know, so I will upload this new version to GitHub (I have to refresh how to do that, to be honest... HAHAHAHA!)

Thanks a lot for your help!

PS: I contacted  SpaceHuhn, the DSTIKE Deauther manufacturer, to check if there was a way to turn off the green leds while not charging, but He told me there is no way by software. Apart from this, the device is awesome!

**Comment by RodricBr** [comment]
Truly, I imagine how satisfying it felt not to see compile errors and seeing the project alive hahaha!

Onto the data format complication, I just checked and my code and yours are exactly the same since I downloaded the project from the .zip file you sent me.
I think the issue might in be the header file instead (`WeatherStation.h`), since this is the configuration file. My format is no longer a `?, 09/03/200` after I change a few configurations in this header file I just mentioned but now it prints out as `SEG, 10/03/2001` (SEG would be Monday or MON for short in english)

To summarize, the date is no longer glitched or being cut but now it is just wrong, so my guess would be that there's a missing entry configuration in the file `WeatherStation.h` that is needed for the

---

Also, as a minor addition I was thinking of adding a "set date" function just like how there's a "set clock" function.
It would look something like this:
```
 [BACK]
 CLOCK & WEATHER (This is by default just "CLOCK" but I decided to add a "& WEATHER" string)
 UPDATE CLOCK
 SET CLOCK
>SET TIME
```
And when selecting "SET TIME"...
```
UP/DOWN to set Date
WDAY, day/mon/year
```
There might be a simpler way of doing it or maybe not doing at all since the date should be updated alongside the time.

---

>\> PS: I contacted SpaceHuhn, the DSTIKE Deauther manufacturer, to check if there was a way to turn off the green leds while not charging, but He told me there is no way by software. Apart from this, the device is awesome!

About the LED , the only thing I dislike is when I select scan and the white LED turns on. As I recall, when I bought the watch a blue light would shine instead. But yes, all and all I definitely agree that the device is pretty awesome!

**Comment by Kalmat** [comment]
Thanks a lot for your help!

Regarding your idea about manually setting the date, I think it would be a fantastic addition. Will you work on it? I strongly encourage you to fork the project, develop that new feature and merge it to the main repo. That way you can learn a lot more about Arduino and C++! (just one comment: the option should be named "SET DATE", do you agree?)

About the wrong date issue, it is very strange to me the output you get. If I understood well, SEG is correct, but not the day of month, the month and the year, right? That's weird since the day of the week is based on all other values.

I had to add 1 to the month (perhaps that is wrong in your case). I also had to adjust the year values adding and substracting 1900 (NTP uses 1900 as base year, then passes it to 1970), 2000 (RTC uses just two digits for the year), etc... A mess.

Perhaps one of the problems is here (line 249, within setTimeRTCfromNTP function) :

`  if (clockYear > 1970) {
`

EDIT: Sorry because I am slowly recalling the structure of the program. The problem can also be here (line 180).

```
    if (firstTimeUpdate) {
      configTime(TZ_SEC, DST_SEC, "pool.ntp.org");
      uint8_t counter = 0;
      while (clockYear <= 2000 && counter < 10) {
        Serial.println("Waiting for NTP server reply... " + String(counter) + " " + String(clockYear));
        delay(500);
        getTimeNTP();
        counter++;
      }
      #ifdef RTC_DS3231
        setTimeRTCfromNTP();
      #endif // #ifdef RTC_DS3231
    }
```

If you are interested, you can debug the returned data from those functions (mainly getTimeNTP and setTimeRTCfromNTP)... I can explain how to do it, so we can understand what can be wrong at that point.

**Comment by RodricBr** [comment]
Hey, forgive me for the late reply! I was busy with some stuff.
But yes, agreed. I might start working on it then when I get some time on my hands and I could name it "set date" too. I also recently bought a Cardputer from M5Stack which will most differently keep me hooked on!

<br>

> About the wrong date issue

Actually no, "SEG" is also incorrect, the whole date format is not being updated. "SEG" would mean "segunda" which is monday in portuguese, since I edited the `WDAY_NAMES_T` array containing the working day string formats.

---

I managed to understand where the problem is coming from. As you can see in the print below, there's an addition "2023" to myClock(DS3231 module) "getDoW, getDate, getMonth, getYear" function that is bringing the wrong working day, date, month and year in sequence.
<img src="https://github.com/user-attachments/assets/043e63cc-96fe-40a8-9d3b-74ea13003754" width=50%>

Apparently the date here is the date on which I first compiled this code and since then it has been stored and counting. It's been 2 days after the first time I successfully compiled the Weather Station adaptation and it makes sense why it is showing now as "QUA" instead of "SEG". (Monday(SEG), tuesday and wednesday(QUA))
This makes me question why the time would be correct if it also comes from DS3231 library.

As always, somebody had a similar issue and, so this guy says, the ideal option would be to use EEPROM library instead of DS3231.
https://arduino.stackexchange.com/a/95923

I recognize that changing the whole code just to fix the date would be kinda of a pain to do but if you have any ideas on other simpler options I'll be happy to hear!

**Comment by Kalmat** [comment]
Hey! No worries at all!!! This is not urgent, and this is just for fun and knowledge... First off, awesome work debugging the app! 

If I understood well, the problem is:

1. A wrong date was set in the RTC hardware module when you first complied the script
2. This wrong date has never been properly updated nor reset
3. The script was retrieving an indistinguishable "coherent but wrong" data from the RTC module (* see note below)
4. I have to check, but very likely the script doesn't update the time and date when retrieving valid (but, again, wrong) time and date from the module

If this is the issue, we just need to retrieve the right time and date from NTP server (if wifi available, of course), and then update the RTC module (**). We can do this after every boot, right after connecting to the wifi and retrieve the info from NTP server, and despite the time and date stored in the RTC module seems valid or not... Does this make sense to you?

(*) as a note aside, the RTC module will keep the time and date unless it runs out of battery. In that case, it will return 01/01/1970 

(**) The discussion you link seems to correspond to a board which has no wifi, so they have to use the time and date when the script was compiled (which also forces to always have the module powered). Instead, we have the much better alternative of using NTP.

**Comment by Kalmat** [comment]
Confirmed (at 90%, to be honest. HAHAHAHAHA!)

Look at this (line 176):

```
    drawProgress(display, 10, "Updating time...");
    if (firstTimeUpdate) {
      firstTimeUpdate = false;
      Serial.println("Updating Internet Time & Date...");
      configTime(TZ_SEC, DST_SEC, "pool.ntp.org");
      uint8_t counter = 0;
      while (clockYear <= 2000 && counter < 10) {
        Serial.println("Waiting for NTP server reply... " + String(counter) + " " + String(clockYear));
        delay(500);
        getTimeNTP();
        counter++;
      }
      #ifdef RTC_DS3231
        setTimeRTCfromNTP();
      #endif // #ifdef RTC_DS3231
```

In case the year retrieved from RTC module (clockYear) seems valid (> 2000), it will not connect to NTP server.

Inside `setTimeRTCfromNTP()` there is another (useless, I think) connection to NTP (using `getTimeNTP()`):

```
void setTimeRTCfromNTP() {
  getTimeNTP();
  if (clockYear > 1970) {
    myClock.setHour(clockHour);
    myClock.setMinute(clockMinute);
    myClock.setSecond(clockSecond);
    myClock.setDate(clockDate);
    myClock.setMonth(clockMonth);
    myClock.setYear(clockYear - 2000);
    myClock.setDoW(clockDoW);
  }
}
```

This should be enough to update the RTC module from NTP data in every boot (`setTimeRTCfromNTP()` is used at that point only), but it is not!

NTP won't answer very quickly after connecting to wifi (don't know why), this is the reason why I had to enclose it within a "while" loop. So, the first time it runs, and in case RTC module has a valid year value, it will too quickly try to retrieve time from NTP, without trying to retry at any point... so the time and date will not be updated.

**Comment by Kalmat** [comment]
Confirmed (at 90%, to be honest. HAHAHAHAHA!)

In case `clockYear` (previously retrieved from RTC module) seems valid (> 2000), the script will not try to update time and date from NTP (using `getTimeNTP() `method). Check this (line 176):

```
    drawProgress(display, 10, "Updating time...");
    if (firstTimeUpdate) {
      firstTimeUpdate = false;
      Serial.println("Updating Internet Time & Date...");
      configTime(TZ_SEC, DST_SEC, "pool.ntp.org");
      uint8_t counter = 0;
      while (clockYear <= 2000 && counter < 10) {
        Serial.println("Waiting for NTP server reply... " + String(counter) + " " + String(clockYear));
        delay(500);
        getTimeNTP();
        counter++;
      }
      #ifdef RTC_DS3231
        setTimeRTCfromNTP();
      #endif // #ifdef RTC_DS3231
    }
```

There is another (useless, I think). call to `getTimeNTP()` within `setTimeRTCfromNTP()` (perhaps I used this last function in another place too, but now it's invoked only once after boot):

```
void setTimeRTCfromNTP() {
  getTimeNTP();
  if (clockYear > 1970) {
    myClock.setHour(clockHour);
    myClock.setMinute(clockMinute);
    myClock.setSecond(clockSecond);
    myClock.setDate(clockDate);
    myClock.setMonth(clockMonth);
    myClock.setYear(clockYear - 2000);
    myClock.setDoW(clockDoW);
  }
}
```

For some reason, NTP is timing out when queried too quickly after connecting to wifi. Now I remember this is the reason why I had to enclose it within a "while" loop (also notice that the delay is placed before the call).

In short, the first time the script runs, it doesn't retrieve the NTP data because year seems valid. Right afterwards, it tries to update within  `setTimeRTCfromNTP()`, with no success; and then it updates RTC module with the very same data the script retrieved before from the module.

I think that forcing to update RTC time and date from NTP on every boot, in case there is wifi available and a valid reply from NTP server will solve the problem.

**Comment by RodricBr** [comment]
Yeah, most definitely seems to be these the problem roots.
The NTP having a timeout/cool down seems to be because the WAF blocks too many request being sent by the same address.

> In short, the first time the script runs [...]

Exactly, and it adds up since the NTP server data never gets fetched in the first boot, the user has to manually go to "update clock" to receive the updated data ("date", "working day", "weather" excluding "time", since it is already stored in the RTC), which is something I noticed while debugging it.
I also think this will solve the problem and, once I can, I'll try to code a little bit to solve that. Again, thank you so much for the help!

Edit: I did in fact configured manually (hardcoded the issue. but hey, it worked! hahaha) the date by adding or subtracting values to the `setTimeRTCfromNTP` function and by doing that it also gave me ideas which I tried many of them and it was very hard to match with the code, so I'll try something different and simpler later.

**Comment by Kalmat** [comment]
Hey! Please, keep me posted whenever you have time to work on the project. No hurries, of course. We all have lots of things to do!

I was thinking that "UPDATE CLOCK" option should also try to fetch time and date from NTP server. At the moment, it only fetches weather data. What do you think? This way the user can update everything as soon as he gets a connection, without the need to manually enter all data, although this manual option is also very convenient.

I would really love you making these changes, so you can learn a bit more about C++ and Arduino (I can help in everything you need, of course); then, merging your code will make you part of the project. To be honest, I love contributions, though unfortunately I had very few in all my projects.
